lib/knetwork/dns.go

```diff
@@ -0,0 +1,191 @@
+package knetwork
+
+import (
+	"errors"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/logger/klog"
+	"github.com/miekg/dns"
+	"net"
+	"strconv"
+	"sync"
+)
+
+var (
+	DNSEntryNotExistError = errors.New("the following entry did not exist")
+	DNSEntryExistError    = errors.New("the following entry did exist")
+)
+
+func NewDNS(mods ...DNSModifier) (*DnsServer, error) {
+	defaultLogger, err := klog.New("default", klog.FromFlags(*klog.DefaultFlags()))
+	if err != nil {
+		return nil, err
+	}
+	s := &DnsServer{
+		routeMap: make(map[string][]string),
+		Logger:   defaultLogger,
+	}
+	for _, mod := range mods {
+		if err := mod(s); err != nil {
+			return nil, err
+		}
+	}
+	return s, nil
+}
+
+type DNSModifier func(s *DnsServer) error
+
+func WithLogger(l *klog.Logger) DNSModifier {
+	return func(s *DnsServer) error {
+		s.Logger = l
+		return nil
+	}
+}
+
+func WithPort(p int) DNSModifier {
+	return func(s *DnsServer) error {
+		s.Port = p
+		return nil
+	}
+}
+
+func WithDomains(domains []string) DNSModifier {
+	return func(s *DnsServer) error {
+		s.domains = domains
+		return nil
+	}
+}
+
+func WithListener(l net.Listener) DNSModifier {
+	return func(s *DnsServer) error {
+		s.Listener = l
+		return nil
+	}
+}
+
+func WithHost(ip string) DNSModifier {
+	return func(s *DnsServer) error {
+		s.host = ip
+		return nil
+	}
+}
+
+type DnsServer struct {
+	dnsServer *dns.Server
+	sync.RWMutex
+	routeMap map[string][]string
+	Port     int
+	Logger   *klog.Logger
+	domains  []string
+	Listener net.Listener
+	host     string
+}
+
+func (s *DnsServer) Start() error {
```

Comment:
`Start()` will block forever, by the fact that it calls ListenAndServe() below? If that's the case, I'd recommend calling it `Run()` instead.

Mild convention: `Start()` starts something and returns. `Run()` loops forever or blocks until completion.